File: src/com/practice/lld/atm/Account.java
package com.practice.lld.atm;

public class Account {

  private final String accountNumber;

  private float balance;

  public Account(String accountNumber, float balance) {
    this.accountNumber = accountNumber;
    this.balance = balance;
  }


  public String getAccountNumber() {
    return accountNumber;
  }

  public float getBalance() {
    return balance;
  }

  public void setBalance(float balance) {
    this.balance = balance;
  }

  public void debit(float amount) {
    this.balance -= amount;
  }

  public void credit(float amount) {
    this.balance += amount;
  }
}

File: go/atm/model/account.go
package model

import (
	"atm/errors"
	"time"
)

type Account struct {
	Number    string
	Ifsc      string
	CreatedAt time.Time
	Cards     map[string]Card
	Balance   float64
}

func (a *Account) Debit(amount float64) error {
	if a.Balance < amount {
		return errors.ErrInsufficientBalance
	}
	a.Balance -= amount
	return nil
}

func (a *Account) Credit(amount float64) error {
	a.Balance += amount
	return nil
}

func NewAccount(number string, ifsc string, createdAt time.Time) *Account {
	return &Account{
		Number:    number,
		Ifsc:      ifsc,
		CreatedAt: createdAt,
	}
}
